### ISSUE TITLE
High thread synchronization overhead

### ISSUE THREAD

**Issue by zerg960** [mapcrafter/mapcrafter#180]
Noticed that there is fairly high synchronization overhead in your implementation - as the number of threads is increased the overall throughput decreases. Here's a table showing my findings, using mapcrafter-gui 2.1 x64 for Windows. Command-line had similar performance as well. I completely delete the output directory before each run. My system has 32 logical processors and ludicrously-fast I/O.

Threads | Tiles per second | Notes
----|-------|----
4 | 67.1
8 | 93.1
10 | 74.4
12 | 59.9
16 | 42.6
24 | 30.2
32 | 20.7
64 | 12.1 | 9% CPU usage - 2 threads per core
96 | 10.1 | 8% CPU usage - 3 threads per core

During these runs I noticed that a single core had significantly higher CPU usage than the rest - and that CPU usage is spent mostly in kernel mode, presumably waiting on a thread synchronization primitive.

I took a quick look at the source code and found that the ThreadManager::mutex is highly contended by the worker threads. I believe performance may be increased by making these changes to the thread manager and work queue classes:
* The thread manager mutex is gone. I don't believe it's necessary as long as there are proper memory barriers.
* Condition variables are always notified when you add something to a queue; they may act as a memory fence, but I'm not 100% sure.
* ConcurrentQueue's mutex can be removed and replaced with similar constructs as well.
* Alternatively: http://moodycamel.com/blog/2014/a-fast-general-purpose-lock-free-queue-for-c++ - but this is a more extensive change.

Fixing this would allow a 7.5GB map to be rendered under 4 minutes on my system, instead of in 16 minutes. I'm not a C++ developer, unfortunately.

Configuration:
```
output_dir = C:\Users\Myself\Documents\git\KMobile\src\main\webapp\map

[world:kreatious]
input_dir = R:\kreatious\2015-05-09-21k
crop_min_x = -5000
crop_min_z = -5000
crop_max_x = 15000
crop_max_z = 15000

[map:map_kreatious]
name = Kreatious
world = kreatious
render_view = topdown
texture_size = 4
tile_width = 4
png_indexed = true
```

The world being rendered is 7.5 GB

I/O speeds, as measured by CrystalDiskMark, with a queue depth of 1, using 32 threads on 16GB of data; to best represent the I/O pattern used by mapcrafter:
> C:\ sequential read/write 17.3/19.3 GB/s
> C:\ random 4K read/write 1.2/1.0 GB/s
> R:\ sequential read/write 18.5/15.6 GB/s 
> R:\ random 4k read/write 1.2/1.0 GB/s

A single thread can sequentially read/write 3.6/3.9 GB/s, and randomly read/write 414.7/440.4 MB/s
Sanity check: `time cat /r/kreatious/2015-05-09-21k/region/*.mca > /dev/null` takes 2.343s

System specifications:
> Windows 10 64-bit
> Running http://mapcrafter.org/windows/main/mapcrafter_2.1_win64.zip using mapcrafter-gui
> 2 Xeon E5-2687W processors (32 logical cores)
> 128GB of RAM
> DIsk read/writes are cached in memory

**Comment by m0r13** [comment]
Wow, thanks for that observation. Unfortunately I don't know that much about multithreading, but I get that your general idea is that I can speed up the multithreading by sparing some of the mutexes, so that threads don't have to wait unnecessarily long for each other?

Can you build Mapcrafter from source and run it on your machine (maybe on Linux)? I would like to test something. You can increase the work that has to be done per item in the queue by increasing the zoom level of the base tiles the worker get to render: https://github.com/mapcrafter/mapcrafter/blob/master/src/mapcraftercore/thread/impl/multithreading.cpp#L125

Set this number to 3 or 4.

**Comment by zerg960** [comment]
Yep - basically, synchronization kills concurrency. If you can get away without synchronizing, you can have more concurrency. Synchronization is needed to protect other threads from seeing an invalid intermediate state.

Results using Ubuntu 14.04 LTS on a brand new virtual machine with dynamic ram & 32 virtual processors:

Threads | Tiles/s with 2 | With 3 | With 4
--------------|---------|-------|-----
8 | 88.70 | 95.16 | 103.10
32 | 57.00 | 42.19 | 35.36

CPU usage with Linux is higher, 

FYI: Mutexes are more expensive in Windows... (but critical sections aren't nearly as expensive: see http://preshing.com/20111124/always-use-a-lightweight-mutex/). Least expensive of all is simply not having them, but then you become vulnerable to multi-threading bugs.



Interesting idea to consider (for a 1 dimensional map):
```java
// Work queues implemented so that each thread gets its own queue...
// and only when a thread's queue is empty will synchronization occur and a thread
// steals half the work from another thread's queue. Watch out for multithreading bugs!
private final ExecutorService threadPool = Executors.newFixedThreadPool(32);

void render(int x, int zoom) {
  if (zoom > maxZoom) {
    maxZoomRender(x);
    return;
  }

  // Combiner can use volatile fields and/or interlocked instructions to avoid the overhead
  // of a memory fence.. since what needs to be synchronized is 1 word long and is
  // independent of all other threads. When both flags are complete, the combiner
  // immediately combines the results on the same thread that last called complete.
  Combiner combiner = new Combiner().whenBothComplete(this::renderCombinedTile());
  threadPool.submit(() -> { render(x * 2, zoom + 1); combiner.complete(0) } );
  threadPool.submit(() -> { render(x * 2 + 1, zoom + 1); combiner.complete(1) } );
}
```
Above code avoids using the more expensive synchronization primitives as much as possible, and keeps the number of highly contended objects to a minimum (the work queues).

There's no need for a global work queue since all that matters is that all tiles get rendered, the order that they get rendered in doesn't matter.

**Comment by zerg960** [comment]
mutrace results with the original code, on a small subset of the original map, using 96 threads:

```
showing statistics for process mapcrafter
mutrace: 7 mutexes used

Mutex #3 first referenced by ...ThreadManager7addWorkERKNS_8renderer10RenderWorkE
Mutex #2 first referenced by ...ThreadManager9getResultERNS_8renderer16RenderWorkResultE
Mutex #1 first referenced by: ...ThreadManager7getWorkERNS_8renderer10RenderWorkE
Mutex #0 first referenced by: ...ConcurrentQueueINS_8renderer10RenderWorkEE4pushES3

mutrace: Showing 4 most contended mutexes:
```

Mutex # | locked | changed | cont | total time [ms] | avg. time [ms] | max time [ms]
----|----|----|----|----|----|----
3 | 1865 | 909 | 33 | 2804.752 | 1.480 | 320.021
2 | 1661 | 648 | 0 | 1088.714 | 0.655 | 320.013
1 | 689 | 530 | 0 | 152.630 | 0.222 | 36.847
0 | 1246 | 440 | 0 | 297.253 | 0.239 | 15.067

Locked: # of times the mutex was acquired
Changed: # of times the owning thread of the mutex changed
Cont: # of times another thread was forced to wait for the mutex
Total time: Total time the lock was locked
Avg time: Average time the lock was held
Max time: Maximum time the lock was held

From this, more than 1% of the runtime was spent holding a lock. Mutexes 3 and 2 are held for a maximum of a third of a second - a fairly wide opportunity for another thread to become blocked. I am not sure as to what causes a majority of the significant slowdown experienced at high thread counts - apparently it's not mutexes. I'll have to do some more analysis to determine the cause, but be aware: I'm not that comfortable with C++.

**Comment by zerg960** [comment]
`valgrind --tool=dtd --trace-cond=yes --trace-mutex=yes --trace-fork-join=yes --exclusive-threshold=10 mapcrafter -c short.conf -j 8 -f map_kreatious` reports that ocassionally that the mutex first seen in `ThreadManager::addWork`, is occasionally held for several hundred milliseconds by `ThreadManager::workFinished`.

Tracing the usage of condition variables it seems that only 1 thread is blocked waiting for signals - the one displaying the progress bar - and the remaining threads merely signal. The frequency that these condition variables are signaled isn't anywhere significant enough to explain the drop in performance (tested using 8 threads, tileSize = 1, textureSize = 1, original code)

I did notice that my configuration for Linux caused I/O wait to be high for the previous comments with the tables - relying on the host OS's IO cache to satisfy requests. I reconfigured to use the guest OS's RAM for I/O. Now the original code with 8 threads finishes in 848s, and 32 threads finishes in 493s (theoretical is 212s); I'll update this post with a revised table. I'd also be willing to allow you root access to the virtual machine if it'll help.

**Comment by m0r13** [comment]
I have to say I'm a little overwhelmed by all those statistics. I need to invest some time to get more into this multithreading stuff. Unfortunately university started again so I'm a bit more busy with other things. Also first I wanted to focus on finishing another feature (the new overlays) first.

How much does this synchronization problem affect performance with a "normal" count of threads (2, 4, 8) ? So maybe we can postpone this problem a bit?

**Comment by zerg960** [comment]
I did a least squares regression analysis on the data using Amdahl's law. It's not a perfect fit since the serial portion varies with the thread count, but it's close (R = 0.99681).

According to the data samples, the portion of the currently implemented algorithm on Linux that is serial is around 1.861% (Windows is going to be higher so performance is worse there - hence why I'm here).

If you are able to get that down to 0.1% on Linux, then you'll get around 2% improvement with 8 threads, 14% improvement with 16 threads, 54% with 24 threads, 231% with 32 threads, 498% with 64 threads, and 1194% with 128 threads.

Common (i7) Linux machines probably won't notice a 2% improvement. Common Windows machines though? They'll notice, more so if they have an SSD. More common processors that'll notice an improvement include the higher end AMD processors and i7 processors. Plus, with higher queue depth, disk drives are able to more efficiently perform I/O. One cheap way to do that is to double the number of threads compared to CPU cores.


As far as learning this multithreading stuff goes, here's some links:
http://kukuruku.co/hub/cpp/lock-free-data-structures-basics-atomicity-and-atomic-primitives
http://g.oswego.edu/dl/jmm/cookbook.html (good for low level conceptual understanding)
http://info.prelert.com/blog/cpp11-mutex-implementations
http://en.cppreference.com/w/cpp/atomic
https://www.arangodb.com/2015/02/comparing-atomic-mutex-rwlocks/
